File: backend/providers/dexscreener.py
import os
import requests
import logging

logger = logging.getLogger("tokencare-dexscreener")
REQUEST_TIMEOUT = 4.0

def fetch_dexscreener_token(contract_address: str, chain_id: str) -> dict:
    """
    Fetches token data from DexScreener API
    """
    api_key = os.getenv("DEXSCREENER_API_KEY", "")
    headers = {}
    if api_key:
        headers["X-API-KEY"] = api_key

    try:
        url = f"https://api.dexscreener.com/latest/dex/tokens/{contract_address.lower()}"
        resp = requests.get(url, headers=headers, timeout=REQUEST_TIMEOUT)
        if resp.status_code == 200:
            data = resp.json()
            pairs = data.get("pairs")
            if pairs and len(pairs) > 0:
                best_pair = pairs[0]
                for p in pairs:
                    if str(p.get("chainId", "")).lower() == str(chain_id).lower():
                        best_pair = p
                        break

                base_token = best_pair.get("baseToken", {})
                price_usd = float(best_pair.get("priceUsd") or 0.0)
                price_change = float((best_pair.get("priceChange") or {}).get("h24") or 0.0)
                volume_24h = float((best_pair.get("volume") or {}).get("h24") or 0.0)
                liquidity_usd = float((best_pair.get("liquidity") or {}).get("usd") or 0.0)
                market_cap = float(best_pair.get("fdv") or best_pair.get("marketCap") or 0.0)

                info = best_pair.get("info", {})
                logo = info.get("imageUrl") or None

                return {
                    "source": "dexscreener",
                    "found": True,
                    "name": base_token.get("name"),
                    "symbol": str(base_token.get("symbol") or "").upper(),
                    "logo": logo,
                    "price": price_usd,
                    "priceUsd": price_usd,
                    "priceChange24h": price_change,
                    "volume24h": volume_24h,
                    "liquidity": liquidity_usd,
                    "marketCap": market_cap,
                    "verified": True
                }
    except Exception as e:
        logger.warning(f"[DexScreener] Error fetching {contract_address}: {e}")

    return {"source": "dexscreener", "found": False}

File: backend/providers/dexscreener.ts
import { NormalizedTokenDetails } from '../types/tokenDetails';

export const CHAIN_MAP_DEXSCREENER: Record<string, string> = {
  '1': 'ethereum',
  '137': 'polygon',
  '56': 'bsc',
  '42161': 'arbitrum',
  '10': 'optimism',
  '8453': 'base',
  '43114': 'avalanche',
  'solana': 'solana',
  'polygon': 'polygon',
  'ethereum': 'ethereum',
  'bsc': 'bsc',
  'arbitrum': 'arbitrum',
  'optimism': 'optimism',
  'base': 'base',
  'avalanche': 'avalanche',
};

export async function fetchDexScreenerToken(
  chain: string,
  contractAddress: string
): Promise<NormalizedTokenDetails | null> {
  const cleanAddr = contractAddress.trim().toLowerCase();
  if (!cleanAddr) return null;

  try {
    const chainKey = CHAIN_MAP_DEXSCREENER[chain.toLowerCase()] || chain.toLowerCase();
    
    // DexScreener token-pairs endpoint
    const url = `https://api.dexscreener.com/latest/dex/tokens/${cleanAddr}`;
    const res = await fetch(url, {
      headers: { 'Accept': 'application/json' },
    });

    if (!res.ok) return null;
    const json = await res.json();
    const pairs = json.pairs;

    if (!pairs || !Array.isArray(pairs) || pairs.length === 0) {
      return null;
    }

    // Filter matching chain if specific, or pick highest liquidity pair
    const matchingPairs = pairs.filter((p: any) => {
      const pChain = (p.chainId || '').toLowerCase();
      return !chainKey || pChain.includes(chainKey) || chainKey.includes(pChain);
    });

    const activePairs = matchingPairs.length > 0 ? matchingPairs : pairs;

    // Sort by USD liquidity descending to get the most canonical pair
    activePairs.sort((a: any, b: any) => {
      const liqA = Number(a?.liquidity?.usd || 0);
      const liqB = Number(b?.liquidity?.usd || 0);
      return liqB - liqA;
    });

    const bestPair = activePairs[0];
    const baseToken = bestPair.baseToken;
    const isBase = (baseToken?.address || '').toLowerCase() === cleanAddr;
    const targetToken = isBase ? baseToken : bestPair.quoteToken;

    // Social and logo extraction
    const info = bestPair.info || {};
    const websites = Array.isArray(info.websites) ? info.websites : [];
    const socials = Array.isArray(info.socials) ? info.socials : [];

    const telegram = socials.find((s: any) => s.type === 'telegram')?.url;
    const twitter = socials.find((s: any) => s.type === 'twitter')?.url;
    const discord = socials.find((s: any) => s.type === 'discord')?.url;
    const website = websites[0]?.url;

    // Determine blockchain standard
    const isSolana = (bestPair.chainId || '').toLowerCase() === 'solana' || chain.toLowerCase() === 'solana';
    const isBsc = (bestPair.chainId || '').toLowerCase() === 'bsc' || chain.toLowerCase() === '56';

    const normalized: NormalizedTokenDetails = {
      name: targetToken?.name || 'Unknown Token',
      symbol: (targetToken?.symbol || 'TOKEN').toUpperCase(),
      decimals: isSolana ? 9 : 18,
      contractAddress: targetToken?.address || cleanAddr,
      chain: bestPair.chainId || chain,
      chainId: chain,
      blockchain: isSolana ? 'Solana' : 'EVM',
      assetStandard: isSolana ? 'SPL' : isBsc ? 'BEP20' : 'ERC20',
      logoUrl: info.imageUrl || undefined,
      bannerUrl: info.header || undefined,
      websiteUrl: website,
      telegramUrl: telegram,
      twitterUrl: twitter,
      discordUrl: discord,
      priceUsd: Number(bestPair.priceUsd) || (isBase ? Number(bestPair.priceUsd) : undefined),
      liquidityUsd: Number(bestPair.liquidity?.usd) || 0,
      fdvUsd: Number(bestPair.fdv) || undefined,
      marketCapUsd: Number(bestPair.marketCap) || Number(bestPair.fdv) || undefined,
      volume24hUsd: Number(bestPair.volume?.h24) || 0,
      priceChange24h: Number(bestPair.priceChange?.h24) || 0,
      pairAddress: bestPair.pairAddress,
      dexName: bestPair.dexId,
      resolvedVia: 'dexscreener',
    };

    return normalized;
  } catch (err) {
    console.error('[DexScreener Provider Error]:', err);
    return null;
  }
}
